### PR TITLE
Builder

### PR DESCRIPTION
**Pull request by bretambrose**
Websocket and proxy support

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    }
+
+    /**
+     * Create a new builder from a Pkcs12 blob
+     *
+     * @param pkcs12Path - The path to a PKCS12 file
+     * @param pkcs12Password  - The password to the PKCS12 file
+     */
+    public static AwsIotMqttConnectionBuilder newMtlsPkcs12Builder(String pkcs12Path, String pkcs12Password) {
+        try (TlsContextOptions tlsContextOptions = TlsContextOptions.createWithMtlsPkcs12(pkcs12Path, pkcs12Password)) {
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java]: Will remove.

File: sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+    protected boolean canReleaseReferencesImmediately() { return true; }
+
+
+    /**
+     * Create a new builder with mTLS file paths
+     * 
+     * @param certPath       - Path to certificate, in PEM format
+     * @param privateKeyPath - Path to private key, in PEM format
+     */
+    public static AwsIotMqttConnectionBuilder newMtlsBuilderFromPath(String certPath, String privateKeyPath) {
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java]: Added a newDefaultBuilder that uses TlsContextOptions.createDefaultClient()

File: samples/BasicPubSub/src/main/java/pubsub/PubSub.java [hunk 3]
```diff
@@ -130,35 +150,63 @@ static void onRejectedError(RejectedError error) {
...
-                MqttClient client = new MqttClient(clientBootstrap, tlsContext);
-                MqttClientConnection connection = new MqttClientConnection(client, new MqttClientConnectionEvents() {
-                    @Override
-                    public void onConnectionInterrupted(int errorCode) {
-                        if (errorCode != 0) {
-                            System.out.println("Connection interrupted: " + errorCode + ": " + CRT.awsErrorString(errorCode));
-                        }
-                    }
+        if (!useWebsockets) {
+            if (certPath == null || keyPath == null) {
```
Comment (justinboswell) [inline samples/BasicPubSub/src/main/java/pubsub/PubSub.java]: Maybe ```System.err.println("If websockets are used, MTLS is required")``` or similar?

File: samples/BasicPubSub/src/main/java/pubsub/PubSub.java [hunk 4]
```diff
@@ -55,7 +59,10 @@ static void printUsage() {
                 "  -k|--key      Path to the IoT thing public key\n"+
                 "  -t|--topic    Topic to subscribe/publish to (optional)\n"+
                 "  -m|--message  Message to publish (optional)\n"+
-                "  -n|--count    Number of messages to publish (optional)"
+                "  -n|--count    Number of messages to publish (optional)\n" +
+                "  -w|--websockets Use websockets\n" +
+                "  --proxyhost   Websocket proxy host to use\n" +
+                "  --proxyport   Websocket proxy port to use\n"
```
Comment (JonathanHenson) [inline samples/BasicPubSub/src/main/java/pubsub/PubSub.java]: add signingRegion

File: samples/BasicPubSub/src/main/java/pubsub/PubSub.java [hunk 5]
```diff
@@ -130,35 +150,63 @@ static void onRejectedError(RejectedError error) {
...
+
+            builder.withBootstrap(clientBootstrap)
+                .withConnectionEventCallbacks(callbacks)
+                .withClientId(clientId)
+                .withEndpoint(endpoint)
+                .withCleanSession(true);
+
+            if (useWebsockets) {
+                builder.withWebsockets(true);
+                builder.withWebsocketSigningRegion("us-east-1");
```
Comment (JonathanHenson) [inline samples/BasicPubSub/src/main/java/pubsub/PubSub.java]: dont hard code this

File: samples/PubSubStress/src/main/java/pubsubstress/PubSubStress.java [hunk 6]
```diff
@@ -299,64 +315,86 @@ private static void cleanupConnections() {
...
+                builder.withCertificateAuthorityFromPath(null, rootCaPath)
+                    .withEndpoint(endpoint)
+                    .withCleanSession(true)
+                    .withBootstrap(clientBootstrap);
 
-                    try {
-                        initConnections(client);
+                if (useWebsockets) {
+                    builder.withWebsockets(true);
+                    builder.withWebsocketSigningRegion("us-east-1");
```
Comment (JonathanHenson) [inline samples/PubSubStress/src/main/java/pubsubstress/PubSubStress.java]: move to cli arg

**Pull request merged**